### ISSUE TITLE
Should thresholds be able to "un-taint" a test?

### ISSUE THREAD

**Issue by uppfinnarn** [loadimpact/k6#107]
Let's say you have this threshold:

```js
export let options = {
    thresholds: {
        http_req_duration: ['avg<=100'],
    },
};
```

Right now, if the test briefly dips below that 100ms average, then recovers for the remainder of the test, it will still be marked as tainted. Should this be the case? Or should the taint status be a reflection of the current state?

**Comment by aidylewis** [comment]
Shouldn't the thresholds be calculated post-test run?

**Comment by uppfinnarn** [comment]
As I remember it, @ragnarlonn really wanted a flag that aborts the test when a threshold fails… not sure if the reasoning for that still stands, it was a good while ago.

The only other thing running them inline could be useful for is showing pass/fail status live in the web dashboard.

**Comment by aidylewis** [comment]
We'd expect web performance to have high outliers which could abort your test almost immediately.

**Comment by uppfinnarn** [comment]
Yeah, that's about what I said too… it was something about long-running tests?

**Comment by aidylewis** [comment]
Soak tests would be normally run at night, expecting results the next day. Or just keep one eye on the real-time graphs.

**Comment by ragnarlonn** [comment]
TL;DR - I suggest we do what @uppfinnarn mentions above - we let tests "recover" and perhaps remove the --abort-on-taint option. Would be a cleaner solution. Tests will run until the end (or they are aborted), and the threshold status at the time the test ends is what determines if the whole test is a pass or a fail.

The reason I wanted to be able to abort tests early if some threshold failed was that I wanted to run many small load tests in a CI test suite, and then you want to save time if possible. If I can abort the test prematurely because my failure criteria has already been met, then that can be useful to shorten the time it takes to run a test suite with many small load tests in it.

Having said that, I think this (being able to abort early) is "luxury" functionality that we really don't know if anyone but me want/need. As @aidylewis points out, looking at the thresholds once after the test is done is really the MVP here - maybe we should focus on that?

One issue also, with checking thresholds during the test, is that you have no control over what set of samples k6 will be using to check the threshold - i.e. I think there is a "grace period" right now, at the start of the test where thresholds aren't checked, right?  Can't remember exactly, but the point is:

- if we check e.g. "95th percentile RTT should be <100ms" and do it after the very first HTTP request, then the test may fail just because that single request happened to be over 100ms (which is not unlikely, given that time-consuming things like DNS lookups tend to happen early in the test).

Introducing a magical grace period, that the user has no means of affecting, is not the best solution. Neither is clumping together samples in sets, maybe based on time intervals, and checking thresholds against samples in a set. We don't know if the intervals/set size is what the user wants. IMO, the only really "clean" solution, that doesn't mean introducing new complex functionality to let people control when/for what data thresholds are checked, is to only check them once after the test has ended - on all samples.

**Comment by uppfinnarn** [comment]
Okay so, thresholds are now the only thing that can taint a test, IsTainted() is updated every 2s. Metrics now report whether they're tainted, and I feel like we may as well keep updating thresholds like this for API querying purposes.
